compiler/rustc_trait_selection/src/solve/fulfill.rs

```diff
@@ -32,11 +32,34 @@ pub struct FulfillmentCtxt<'tcx> {
     /// gets rolled back. Because of this we explicitly check that we only
     /// use the context in exactly this snapshot.
     usable_in_snapshot: usize,
+
+    track_obligations: bool,
 }
 
 impl<'tcx> FulfillmentCtxt<'tcx> {
     pub fn new(infcx: &InferCtxt<'tcx>) -> FulfillmentCtxt<'tcx> {
-        FulfillmentCtxt { obligations: Vec::new(), usable_in_snapshot: infcx.num_open_snapshots() }
+        FulfillmentCtxt {
+            obligations: Vec::new(),
+            usable_in_snapshot: infcx.num_open_snapshots(),
+            track_obligations: infcx.tcx.sess.opts.unstable_opts.track_trait_obligations,
+        }
+    }
+
+    fn track_evaluated_obligation(
+        &self,
+        infcx: &InferCtxt<'tcx>,
+        obligation: &PredicateObligation<'tcx>,
+        result: &Result<(bool, Certainty, Vec<Goal<'tcx, ty::Predicate<'tcx>>>), NoSolution>,
+    ) {
+        if self.track_obligations {
+            if let Some(inspector) = infcx.obligation_inspector.get() {
+                let result = match result {
+                    Ok((_, c, _)) => Ok(*c),
+                    Err(NoSolution) => Err(NoSolution),
+                };
+                (inspector)(infcx, &obligation, result);
+            }
```

Comment:
isn't the `Some(inspector)` check enough :thinking: why did you add a separate flag here?

Reply:
In it's current state, it's technically possible to have a `Some` when the `track-trait-obligation` flag isn't set. I'll change it to check the flag before setting the option.

Reply:
Ah, I see. The motivation to have the unstable flag was to make the changes seem less intrusive :) especially because using the inspector relies on using the next trait solver, which is behind a flag. But functionally there's no reason to have the flag.

Reply:
I removed the unstable flag and added a comment to `inspect_typeck` which states that it only has an effect when using the next trait solver. That should be clear enough for its use. I'm not particularly worried that someone will rely on its behavior during normal compilation.